cpp/include/raft/neighbors/detail/refine.cuh

```diff
@@ -0,0 +1,232 @@
+                 candidates.extent(0) == n_queries,
+               "Number of rows in output indices and distances matrices must equal number of rows "
+               "in search matrix.");
+
+  RAFT_EXPECTS(indices.extent(1) == k,
+               "Number of columns in output indices and distances matrices must be equal to k");
+
+  RAFT_EXPECTS(queries.extent(1) == dataset.extent(1),
+               "Number of columns must be equal for dataset and queries");
+
+  RAFT_EXPECTS(candidates.extent(1) >= k,
+               "Number of neighbor candidates must not be smaller than k (%d vs %d)",
+               static_cast<int>(candidates.extent(1)),
+               static_cast<int>(k));
+}
+
+/**
+ * See raft::neighbors::refine for docs.
+ */
+template <typename idx_t, typename data_t, typename distance_t, typename matrix_idx>
+void refine_device(raft::handle_t const& handle,
+                   raft::device_matrix_view<const data_t, matrix_idx, row_major> dataset,
+                   raft::device_matrix_view<const data_t, matrix_idx, row_major> queries,
+                   raft::device_matrix_view<const idx_t, matrix_idx, row_major> neighbor_candidates,
+                   raft::device_matrix_view<idx_t, matrix_idx, row_major> indices,
+                   raft::device_matrix_view<distance_t, matrix_idx, row_major> distances,
+                   distance::DistanceType metric = distance::DistanceType::L2Unexpanded)
+{
+  matrix_idx n_candidates = neighbor_candidates.extent(1);
+  matrix_idx n_queries    = queries.extent(0);
+  matrix_idx dim          = dataset.extent(1);
+  uint32_t k              = static_cast<uint32_t>(indices.extent(1));
+
+  common::nvtx::range<common::nvtx::domain::raft> fun_scope(
+    "neighbors::refine(%zu, %u)", size_t(n_queries), uint32_t(n_candidates));
+
+  check_input(dataset.extents(),
+              queries.extents(),
+              neighbor_candidates.extents(),
+              indices.extents(),
+              distances.extents(),
+              metric);
+
+  // The refinement search can be mapped to an IVF flat search:
+  // - We consider that the candidate vectors form a cluster, separately for each query.
+  // - In other words, the n_queries * n_candidates vectors form n_queries clusters, each with
+  //   n_candidates elements.
+  // - We consider that the coarse level search is already performed and assigned a single cluster
+  //   to search for each query (the cluster formed from the corresponding candidates).
+  // - We run IVF flat search with n_probes=1 to select the best k elements of the candidates.
+  rmm::device_uvector<uint32_t> fake_coarse_idx(n_queries, handle.get_stream());
+
+  thrust::sequence(
+    handle.get_thrust_policy(), fake_coarse_idx.data(), fake_coarse_idx.data() + n_queries);
+
+  raft::neighbors::ivf_flat::index<data_t, idx_t> refinement_index(
+    handle, metric, n_queries, false, dim);
+
+  raft::spatial::knn::ivf_flat::detail::fill_refinement_index(handle,
```

Comment:
Just a little note: I'm hoping to remove the raft::spatial::knn namespace altogether in 23.02 and officially move all the detail headers over to raft::neighbors. Hopefully we can get all the deprecated files removed as well.